### PR TITLE
spark-backend CounterPipe

### PR DESCRIPTION
**Pull request by hazel-sudz**
so counterPipe looks to be good... the discussion point I think is the best way to make `addForce` a blocking call.

### CODE REVIEW

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (hazel-sudz) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: I am currently confused why getting the SparkCounters state is empty inside of this map scope but if I compute it lazily it has the correct values later when asserting in tests.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: when placing breakpoints, the map block above hits before the counterPipe is processed in spark

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: @johnynek any tips?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala]: perf nits:

1. access `counterAction._1` once
2. iterator is better since it is always inlined

```scala
resolve.map { x =>
  val counts = x._2.iterator
  while (counts.hasNext) {
    val ((g, c), amt) = counts.next()
    SparkCounters.addToCounter(g, c, amt)
  }
  x._1
}
```

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: 1. can we make this private?
2. I would probably instead do `@transient private lazy val accumulator = new AtomicReference[SparkCountersInternal](null)` and use atomic operations since I think you can have multithreading in spark.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: if you do this in a loop you are calling `getAccumulator` many times. I would also add a method that accepts `Iterable[((String, String), Long)]` or something so you can get once, then add everything.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: why make this a val?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: if you make this `AccumulatorV2[((String, String), Long), ...]` can you avoid reboxing the counters?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: `0L` would be better to be more specific.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: I think this is probably better done in `newWriter` in `SparkMode` or in the constructor here.

You can have many stages to an `Execution`. Do you want to do this on every stage? I don't think so.

Also, I think ideally we could have a cleanup where maybe we unregister when the Writer is closed?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: These futures complete when spark has returned an RDD in many cases. That may not be when the job is finished.

That's an issue. That hasn't been a problem that I know of so far. So we really need to make sure this Future is only complete when *all* of these writes have fully occurred.

I think the problem is probably with `def force`. So, this part of the future should occur earlier:

https://github.com/twitter/scalding/blob/e44f2f21fa90f669d6bf81af607c5d537f19e93c/scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala#L90

and in the case of `forceToDiskExecution` we need some way for spark to tell us when the caching/materialization is done. I don't know how to do that off the top of my head.

Alternatively, rather than using cache, we could make a temp file ourselves, write to that, which is a spark action and we can block on that, then we can re-read that path ourselves.

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala
Comment (johnynek) [inline scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala]: we are using `toIterableExecution` here, so the fact that the `toLocalIterator` is being done in the wrong place is likely your issue.

**Comment by hazel-sudz** [comment]
thank you so much! I will look into making the changes

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: > These futures complete when spark has returned an RDD in many cases. That may not be when the job is finished.

same conclusion I came to I will look into trying to solve this issue I might also split it off into a separate PR if I figure it out

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: there is a pattern that was historically used inside of spark sql of using the `count` operation to force an RDD computation: 

https://github.com/apache/spark/pull/30403/files#diff-aa467582590608ec702c1caf6208e79af5e6f3ee2cfd58a53bdf2601cd5015d6R65

this is one way to achieve a blocking `cache` operation though it does carry some minimal overhead of re-reading the RDD. 

it's surprising that there's a blocking version of `unpersist` but no blocking `persist` on RDD

Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: okay, that makes sense for the Force case, but I think being more direct and fixing the bug with toIterable.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala]: done

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (hazel-sudz) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: okay I currently have this hack to make the addForce method be blocking. I'm sure there is a better way but I currently can't think of one. Is this alright for now?

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala
Comment (hazel-sudz) [reply scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala]: I believe the call to toLocalIterator happens after the `execute` call for `SparkWriter`. So I'm not sure how we can lift that call for `execute` to be aware of it. 

But... that would definitely be the cleanest solution to use instead of the `count` hack I currently have

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala
Comment (hazel-sudz) [inline scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala]: added a more tricky test case

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: done

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: moved into SparkWriter constructor and fixed for it to be compatible with multistage execution.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: done

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: done

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: fixed

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: I don't think 2 is necessary. Check out this explanation by someone familiar with spark internals: 
https://stackoverflow.com/a/43912747

> Accumulators are not thread-safe, in fact they do not need to be thread-safe. For executors, accumulators are write only variables, they can be added by executors and they can be read by the driver. Driver makes use of DAGScheduler.updateAccumulators method to update values of accumulators after task is completed, and this method is called only from a thread that runs scheduling loop. At a time, only one task completion event is handled. That's why there is no need for accumulators to be thread-safe.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: for 1 I changed the implementation so not sure if it's still applicable

**Comment by johnynek** [comment]
can you merge develop now that I've merged the format?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: let's remove the `{ }` please.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: can this be private? Can you add a comment on why this mutable value is safe? Is it never used in a multi-threaded context?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: Can we remove the side-effecting default value. Instead can you make a method on the companion:

```scala
object SparkCounters {
  def withRandomContextPrefix(sparkSession: SparkSession): SparkCounters =
    new SparkCounters(sparkSession, java.util.UUID.randomUUID.toString)
}
```

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: why doesn't `def keys = copyState.keySet` work here? copystate never changes after this method right?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: unfortunately, `getOrElseUpdate` allocates since the `else` clause is lazy. Can we move to:

```scala
val prev =
  mapState.get(count._1) match {
    case Some(v) => v
    case None => 0L
  }
mapState(count._1) = prev + count._2
```

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: convention in scala is to use `()` for side-effecting things, like this: `val count = in.next()`

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: same comment as above about using the match pattern.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: mapState is already a map, so I don't think you need this `getOrElseUpdate`. I think:

```scala
val bldr = MutableMap.newBuilder[(String, String), Long]()
bldr ++= mapState
bldr.result()
```
is the more efficient way to do this.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: I think it is better to do:
```scala
val prev = mapState.get(kvPair._1) {
  case Some(v) => v
  case None => 0L
}
mapState(kvPair._1) = prev + kvPair._2
```

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: shouldn't we get the `forcedRdd` from the `WorkVal` you have here? I think so. If not, please add a comment why not.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: I think we actually want:

```scala
forcedRdd.flatMap { x =>
  Future {
    scala.concurrent.blocking {
      x.count()
      ()
    }
  }
}
```

Because we don't want to block the mapping thread. (we should review that we are using `scala.concurrent.blocking` on blocking operations.

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala
Comment (johnynek) [inline scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala]: do you actually need the `Ordering` here? I don't see why.

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala
Comment (johnynek) [inline scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBackendTests.scala]: Can we also do a version where you write to a null-sink or text file or something and make sure the `writeExecution` version also works?

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: this isn't thread-safe: mapState can change while iterating.

Can you comment why this is safe?

Alternatives:
1. use `mapState.synchronized { }` before reading/writing
2. use an `immutable.Map[(String, String), Long]` and an `AtomicReference` so taking a copy is very cheap (although updates will be maybe twice as slow, so maybe the mutex approach above is better).

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: all of these methods on accumulator operate on a "share nothing" model. The driver maintains the "true" state of the accumulator and merges updates from other accumulator states returned from executors. I am 99% sure that the merging process does not require thread safety to function. 

If you look at the implementation for what is passed to executors: 
```scala
  // Called by Java when serializing an object
  final protected def writeReplace(): Any = {
    if (atDriverSide) {
      if (!isRegistered) {
        throw new UnsupportedOperationException(
          "Accumulator must be registered before send to executor")
      }
      val copyAcc = copyAndReset()
      assert(copyAcc.isZero, "copyAndReset must return a zero value copy")
      val isInternalAcc = name.isDefined && name.get.startsWith(InternalAccumulator.METRICS_PREFIX)
      if (isInternalAcc) {
        // Do not serialize the name of internal accumulator and send it to executor.
        copyAcc.metadata = metadata.copy(name = None)
      } else {
        // For non-internal accumulators, we still need to send the name because users may need to
        // access the accumulator name at executor side, or they may keep the accumulators sent from
        // executors and access the name when the registered accumulator is already garbage
        // collected(e.g. SQLMetrics).
        copyAcc.metadata = metadata
      }
      copyAcc
    } else {
      this
    }
  }
``` 

executor threads receive a state-less version of the accumulator.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: the only time we read the value of the driver accumulator is at the end after execution has already ended and all updates to the accumulator have been merged by spark.

Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: Okay, can you add that comment to the code around the class something like:

```scala
/**
 * regarding thread-safety: .... 
 * permalink to this code: ... 
```

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: hmmm maybe not I'm starting to doubt based on the official example: https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/util/AccumulatorV2.scala#L451

I want to try to write a unit test to prove if this is necessary because synchronizing threads is going to be a big perf hit.

Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: Sync won't be a big perf hit: you are copying here. The added cost of synchronization will not be large. Also this probably happens very rarely (at some kind of boundary, such as the task finishing). I would be surprised if you could measure a hit.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: good point this only happens after an rdd actions starts/returns. So assuming a very complicated job has ~100-1000 rdd actions this is no problem

 ok let's just do it better safe than sorry

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: changed all methods to be thread safe based on reported crashes and resolutions in https://issues.apache.org/jira/browse/SPARK-17463 after digging for info

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: I wonder if your failing test is due to this comment? You aren't returning a previously forced value.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: I have not looked at that yet. But one other thing I suspect: the counter increment is done through map transform on RDD. Spark does not guarantee counter consistency in transformation (only actions).

is we want strict consistency we would need a blocking action like `forEach`

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: the reason I suspect this is because the result is a strict multiple of the expected value (200 returned vs 100 expected). This leads me to think that the map phase was recomputed twice by the spark engine.

Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: Yes, but it could also be because the planner asked Spark to compute it twice. Not clear to me. I would investigate and try to falsify both.

Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: > shouldn't we get the `forcedRdd` from the `WorkVal` you have here? I think so. If not, please add a comment why not.

I am a bit confused as to what is being referred to here. `WorkVal` is just `val workVal: WorkVal[T] = (src, forcedRdd)` no?.

Anyways it does appear that we have a double computation somewhere in the planner. Specifically happens in the case that this line is included: `.flatMap(_ => TypedPipe.from(sinkExample).toIterableExecution)` after an execution.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (hazel-sudz) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: I believe I located the problem with the failing test. We need to reset the counter state before we return from here. I think what's going on is that flapMapping the execution re-uses the same writer which duplicates the counter values.

I confirmed that the actual counter incrementation occurs only once by stepping through breakpoints and checking on the counter state.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala
Comment (johnynek) [inline scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala]: nit:

```scala
// get this without holding the lock here, so we avoiding holding two locks simultaneously
val thatMapState = other.value
mapState.synchronized {
 for (kvPair <- thatMapState) { ... 
}
```

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: yeah `WorkVal` is that pair, but if you have a cached value, you are ignoring the cached value and returning the input. I think you need to use that cached value. Don't you? If not, can you explain why we can safely ignore an existing `Future[RDD[_ <: A]]` given that we already have another equivalent computation?

I don't see why, but it could be true.

Basically: if we have two identical optimized typedpipes, they should share the exact same `Future[RDD...`.

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: We still need to address this. Otherwise this call may block a compute thread which can really harm concurrency.

File: scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBaseTest.scala
Comment (johnynek) [inline scalding-spark/src/test/scala/com/twitter/scalding/spark_backend/SparkBaseTest.scala]: can we remove `[A]` and use `Execution[Any]`

File: scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala
Comment (hazel-sudz) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: ok yeah if I'm understanding correctly we should do this. Sounds good to me if this is the case: 

```scala
-        case Some(_) =>
-          (copy(initToOpt = initToOpt + ((c, init) -> opt)), false, rdd)
+        case Some(cacheHit) =>
+          (copy(initToOpt = initToOpt + ((c, init) -> opt)), false, cacheHit._2)

```

Comment (johnynek) [reply scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala]: yeah, I think so, or I would write `case Some((_, frdd)) => ...`

**Review by johnynek — APPROVED** [review]

**Comment by johnynek** [comment]
This is great! Thanks.

**Pull request merged**
